app/assets/javascripts/discourse/models/topic.js

```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
     }
   },
 
+  firstPost() {
+    const postStream = this.postStream;
+    let firstPost = postStream.get("posts.firstObject");
+
+    if (firstPost.get("post_number") === 1) {
+      return Promise.resolve(firstPost);
+    }
+
+    const postId = postStream.findPostIdForPostNumber(1);
+
+    // try loading from identity map first
+    firstPost = postStream.findLoadedPost(postId);
+    if (firstPost) {
+      return Promise.resolve(firstPost);
+    }
+
+    return this.postStream.loadPost(postId);
+  },
+
   toggleBookmark() {
     if (this.bookmarking) {
       return Promise.resolve();
     }
     this.set("bookmarking", true);
-
-    const stream = this.postStream;
-    const posts = get(stream, "posts");
-    const firstPost =
-      posts && posts[0] && posts[0].get("post_number") === 1 && posts[0];
     const bookmark = !this.bookmarked;
-
-    const toggleBookmarkOnServer = () => {
-      if (bookmark) {
-        if (this.siteSettings.enable_bookmarks_with_reminders) {
-          return firstPost.toggleBookmarkWithReminder().then(() => {
-            this.set("bookmarking", false);
-            return this.afterTopicBookmarked(firstPost);
-          });
+    let posts = get(this.postStream, "posts");
+
+    return this.firstPost().then(firstPost => {
+      const toggleBookmarkOnServer = () => {
+        if (bookmark) {
+          if (this.siteSettings.enable_bookmarks_with_reminders) {
+            return firstPost.toggleBookmarkWithReminder().then(() => {
+              this.set("bookmarking", false);
+              return this.afterTopicBookmarked(firstPost);
+            });
+          } else {
+            return ajax(`/t/${this.id}/bookmark`, { type: "PUT" })
+              .then(() => {
+                this.toggleProperty("bookmarked");
+                return this.afterTopicBookmarked(firstPost);
+              })
+              .catch(popupAjaxError)
+              .finally(() => this.set("bookmarking", false));
+          }
         } else {
-          return ajax(`/t/${this.id}/bookmark`, { type: "PUT" })
+          return ajax(`/t/${this.id}/remove_bookmarks`, { type: "PUT" })
```

Comment:
I know this is not related to this PR, but shouldn't removing a bookmark be a DELETE request and not a PUT ?

Reply:
Good point, will open another PR to fix this